pkg/controllers/migration/crd/controller.go

```diff
@@ -0,0 +1,99 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package crd
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/awslabs/operatorpkg/object"
+	"github.com/awslabs/operatorpkg/status"
+	"github.com/samber/lo"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	"sigs.k8s.io/karpenter/pkg/apis"
+	v1 "sigs.k8s.io/karpenter/pkg/apis/v1"
+	"sigs.k8s.io/karpenter/pkg/cloudprovider"
+	"sigs.k8s.io/karpenter/pkg/operator/injection"
+)
+
+type Controller struct {
+	cloudProvider cloudprovider.CloudProvider
+	kubeClient    client.Client
+}
+
+func NewController(client client.Client, cloudProvider cloudprovider.CloudProvider) *Controller {
+	return &Controller{
+		cloudProvider: cloudProvider,
+		kubeClient:    client,
+	}
+}
+
+func (c *Controller) Reconcile(ctx context.Context, crd *apiextensionsv1.CustomResourceDefinition) (reconcile.Result, error) {
+	ctx = injection.WithControllerName(ctx, "migration.crd")
+
+	nodeClassGvk := lo.Map(c.cloudProvider.GetSupportedNodeClasses(), func(nc status.Object, _ int) schema.GroupVersionKind {
+		return object.GVK(nc)
+	})[0]
+	if _, ok := lo.Find(apis.CRDs, func(item *apiextensionsv1.CustomResourceDefinition) bool {
+		return item.ObjectMeta.Name == crd.Name
+	}); !ok && nodeClassGvk.Kind != crd.Spec.Names.Kind {
+		return reconcile.Result{}, nil
+	}
+	if !lo.Contains(crd.Status.StoredVersions, "v1") {
+		return reconcile.Result{}, nil
+	}
+	list := &unstructured.UnstructuredList{}
+	list.SetGroupVersionKind(
+		schema.GroupVersionKind{
+			Group:   crd.Spec.Group,
+			Version: "v1",
+			Kind:    crd.Spec.Names.Kind,
+		},
+	)
+	if err := c.kubeClient.List(ctx, list); err != nil {
+		return reconcile.Result{}, fmt.Errorf("listing %s(s), %w", crd.Spec.Names.Kind, err)
+	}
+	for _, item := range list.Items {
+		// requeue if a custom resource is found without annotation
+		if !lo.HasKey(item.GetAnnotations(), v1.StoredVersionMigratedKey) {
+			return reconcile.Result{RequeueAfter: 1 * time.Minute}, nil
+		}
+	}
+	// if all custom resources have been updated, patch the CRD
+	stored := crd.DeepCopy()
+	crd.Status.StoredVersions = []string{"v1"}
+	if err := c.kubeClient.Status().Patch(ctx, crd, client.MergeFromWithOptions(stored, client.MergeFromWithOptimisticLock{})); err != nil {
```

Comment:
Have we tried strategic merges? I know we can't do strategic merge on the CR itself, and I assume this extends to the CRD, but haven't tried.

Reply:
Done.